Make public asset use explicit

schneems opened the pull request:
Currently when an asset isn't found the behavior is to pass the string through. For example a valid asset will return a url from pipeline

``` ruby
asset_path("application.js")
# => assets/application-123098udasvi0mnafd.js
```

While if you make a typo, you won't get an error or anything, it just falls through:

``` ruby
asset_path("app1icati0n.js")
# => "app1icati0n.js"
```

This is bad. No exceptions are raised, the error goes undetected.

This PR makes adds an explicit api for when you want to use an asset that is not in the asset pipeline. If you want an asset from the "public" directory you can now explicitly say so by using `public_folder: true`

``` ruby
asset_path("app1icati0n.js", public_folder: true)
# => "app1icati0n.js"
```

Now there is no confusion by sprockets-rails about the intent of the method call. Likewise in rails/sprockets-rails#375 we are adding a flag `unknown_asset_fallback`. When it is set to `true` then the current fallback behavior will happen, but a deprecation will be emitted. When it is set to false an exception will be raised so you know immediately that the asset was not found.

We are defaulting it to `false` for all new apps.

``` ruby
# Unknown asset fallback will return the path passed in when the given
# asset is not present in the asset pipeline.
Rails.application.config.assets.unknown_asset_fallback = false
```

While this patch is intended to go along with the sprockets-rails patch, it should still function with older sprockets-rails versions, however no deprecations or exceptions will be raised.

schneems committed:
Allow a more explicit public behavior

We want to make it more explicit when a user wants to avoid the asset pipeline to do this we will add `public_*` methods to all path helpers. So if someone wants to use an asset that isn't maintained by the asset pipeline they can use `public_asset_path` instead of `asset_path` and letting it fall through.

The main reason for this change is the desire to raise helpful errors in the future. Right now if you typo an asset name, then we assume you intended an asset in the `public/` folder and not in the pipeline so nothing fails and the error sits silently until you realize the page didn't render correctly. We have to deprecate today so we can raise meaningful error messages in the future.

schneems committed:
Make sure routes don't interfere with paths

schneems committed:
Add descriptive comment

schneems committed:
Add public URL helpers

schneems committed:
Add docs to `public_*_path` methods & fix comments

schneems committed:
Match method signature

schneems committed:
Test `public_` methods.

schneems committed:
Add `public_*` helpers to all the `_tag` methods.

schneems committed:
Fix formatting

schneems commented:
The sprockets-rails PR is https://github.com/rails/sprockets-rails/pull/375

alexcameron89 commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -304,19 +344,27 @@ def audio_tag(*sources)
       end
 
+      # Returns an HTML audio tag for the source asset in the public
+      # folder. This uses +audio_tag+ and skips any asset
```
You skipped this line in the documentation here: `# lookups by assuming any assets are in the `public` folder.`

alexcameron89 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -158,6 +167,14 @@ def asset_path(source, options = {})
       alias_method :path_to_asset, :asset_path # aliased to avoid conflicts with an asset_path named route
 
+      # Computes the path to an asset in the public folder.
+      # This uses +asset_path+ and skips any asset lookups by assuming the asset is in the
```
On these docs, should it be `+path_to_asset+`? Same question for the ones below.

matthewd commented:
This feels too onerous to me. Particularly, having the asset pipeline hijack `image_tag` seems out of step with how other similar tag helpers behave.

schneems commented:
@md not sure I follow. Asset pipeline already hijacks these helpers. This patch provides an un-hijack.

robin850 commented:
:+1: on deprecating but can't we just add the `:public_folder` option to the existing helpers and output the warning if none of `:public_folder` or `:raw` is passed ? Thus we don't have to maintain more methods and more documentation for two things that are mostly the same.

matthewd commented:
@schneems the distinction I'm making is that it currently *enhances* the helper, but doesn't interfere with non-asset use.

This also creates a cascade problem, where because `image_path` becomes `public_image_path`, `image_tag` must become `public_image_tag`, and so must some 3rd party helper that calls `image_tag`, etc.

I propose that we add `public_image_path`, and deprecate calling `image_path` with a **relative** filename that is not a known asset... and that's all.

So, anyone using `image_tag` with an asset, or a full image URL, or `/some/thing.png`, is unaffected -- all of those strings mean exactly one thing, and there's no potential for misinterpretation. Anyone currently using `image_tag('my/image.png')` to refer to something in `public/images/my/image.png` will be warned to change that to: `image_tag(public_image_path('my/image.png'))` -- I feel no obligation to offer a `public_image_tag`, because that formulation is relatively rare. And so the 3rd party helper will also be unaffected: it takes a string in, and the application-level caller introduces the `public_image_path` helper if they're doing the unusual thing.

schneems commented:
I never liked the idea of introducing so many new helpers, method lookup time isn't free and more methods can mean slower performance over time. I do like the idea of just exposing the `public_folder: true` API. It's basically almost already implemented by this patch, and makes intent clear.

The code in rails needs to enable people to run the same app they do today without deprecations. It also needs to be written in such a way that we can detect correctly when we should be deprecating (and eventually raising). I'm not in love with the idea of something like `image_tag(public_image_path('my/image.png'))`, it looks a bit gnarly and the only way I can see to detect when it's not being used correctly is to have `public_image_path` return a special object instead of a string.

I'm going to explore modifying this patch to get rid of the `public_*` helpers, which I think everyone agrees are not the best idea.

## Side Notes

I wanted to note that I don't like the API of passing in an asset that starts with a slash as indicating a full path. I deprecated that. I think it's too easy to accidentally start an asset with a slash and not realize that you're bypassing the whole asset pipeline, silently. At the end of the day I want an error to show up in development if someone uses one of these helpers with anything that won't resolve to an asset that exists and render in production.

I also wanted to note a special case in `asset_path` for the full url option: When you call `image_tag` with something like `image_tag("https://www.schneems.com/smile.png")` the call will eventually call `asset_path` before it calls `compute_asset_path` which is the main entry point to the asset pipeline. Before that can happen

```ruby
def asset_path(source, options = {})
  raise ArgumentError, "nil is not a valid asset source" if source.nil?

  source = source.to_s
  return "" if source.blank?
  return source if URI_REGEXP.match?(source) # <====== HERE
```

This last line will return true since we're giving it a full URI so the asset lookup never happens, no deprecation is thrown. So we don't need to handle that case for skipping deprecations.

Thanks for all then input. Let me know if you need more clarification on any of this, or want to respond to any of the sub-points.

schneems committed:
Favor  `public_folder: true` over `public_*`

Adding all those `public_*` methods is a bit heavy handed, we can change the API to instead  use `public_folder: true`. Change was pretty easy since it was already implemented that way.

matthewd commented:
> I wanted to note that I don't like the API of passing in an asset that starts with a slash as indicating a full path. I deprecated that.

IMO it's no more acceptable for `image_tag('/a/url')` to start insisting on assets than it would be for `link_to('/a/url')` to start complaining about the argument not matching a route: they're both direct convenience wrappers for their corresponding HTML tags, and it's only a secondary / additional convenience that they can intuit a "better" bonus behaviour for some arguments.

> I think it's too easy to accidentally start an asset with a slash and not realize that you're bypassing the whole asset pipeline, silently.

It's not supposed to be; that's why we turned on digests in development. If you use `/assets/my/image.png`, that's already supposed to complain & fail. And if you use `/my/image.png`, it won't complain, but it also won't ever work -- and I could likely be convinced that should raise **if** `my/image.png` is a valid asset.

The only other possibility is that someone manually determines and then uses `/assets/my/image-012345etc.png`. That's going to go pretty poorly, but feels like a reasonable limit-point to me... the gun is braced to point horizontally and down-range: if you still manage to get your foot in front of it, good luck to you.

(Which of those three is the one that worries you? The first certainly used to be a big problem, but that's why we fixed it... or at least I thought we did.)

---

> I'm not in love with the idea of something like image_tag(public_image_path('my/image.png')), it looks a bit gnarly 

My first thought was to just kill the `/images` magic completely... given that the asset pipeline has been the "right" path for a long time now, I'm dubious that it sees much use. And people that *are* using it could just switch to `/images/my/image.png` -- the magic of the asset pipeline seems worthwhile for what it gives you... inserting 8 characters feels a bit more opaque-for-the-sake-of-it.

With that context, the above doesn't trouble me overmuch... the two methods are at very different conceptual levels, just like `link_to(new_post_comment_path(@post))`. Nice API is nice, but we're not here to prevent people from ever having to call more than one method.

> and the only way I can see to detect when it's not being used correctly is to have public_image_path return a special object instead of a string.

It'd be returning an absolute URL `/images/my/image.png`, whereas I only want to complain about a bare `my/image.png`.

> This last line will return true since we're giving it a full URI so the asset lookup never happens, no deprecation is thrown.

:+1:

schneems commented:
I updated to use `public_folder: true` as the API and got rid of the `public_*` methods. Need to update sprockets-rails patch.

schneems commented:
Updated sprockets-rails patch, was pretty minimal.

schneems committed:
Missed 2 public_* methods in cleanup

schneems committed:
Default assets.unknown_asset_fallback to false

schneems committed:
Do not deprecate using assets with leading slash

schneems committed:
Fully document asset_path.

schneems committed:
Fix initializer

schneems committed:
[ci skip] Upgrading assets.unknown_asset_fallback

schneems committed:
[ci skip] asset_pipeline.md assets.unknown_asset_fallback

schneems committed:
[ci skip] configuring.md assets.unknown_asset_fallback

schneems committed:
Thanks rubocop

matthewd commented on railties/lib/rails/generators/rails/app/templates/config/initializers/assets.rb.tt:
```diff
@@ -6,6 +6,10 @@ Rails.application.config.assets.version = '1.0'
 
+# Unknown asset fallback will return the path passed in when the given
+# asset is not present in the asset pipeline.
+Rails.application.config.assets.unknown_asset_fallback = false
```
I think this has to go elsewhere so the upgrade task can skip it :confused:

schneems replied:
Moved to the new initializer

schneems committed:
Move default to different initializer

Is this usage correct? cc/ @prathamesh-sonpatki

prathamesh-sonpatki commented on guides/source/asset_pipeline.md:
```diff
@@ -586,6 +586,19 @@ in your application are included in the `config.assets.precompile` list.
 
+### Raise an Error When an Asset is Not Found
+
+If you are using a recent version of sprockets-rails you can configure what happens
```
may be specify exactly which version of sprockets-rails, like greater than > 3.something ?

schneems committed:
Set unknown_asset_fallback false only new apps

schneems committed:
Specify version of Sprockets-rails this will be in

schneems commented:
Updated to `Rails.application.config.assets.unknown_asset_fallback = <%= options[:update] ? true : false %>` and mentioned the specific version of sprockets-rails in the assets guide,

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      # This is the entry point for all assets.
+      # When using the asset pipeline (i.e. sprockets and sprockets-rails), the
+      # behavior is "enhanced". You can bypass the asset pipeline by passing in
+      # `public_folder: true` to the options.
```
The RDoc parser doesn't understand back-ticks, you should rather use `<tt>public_folder: true</tt>`.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
+      #
+      # === Without the asset pipeline (`public_folder: true`)
```
Ditto, use `<tt>` here.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
-      #   asset_path "http://www.example.com/js/xmlhr.js" # => http://www.example.com/js/xmlhr.js
+      # === With the asset pipeline
+      #
+      # All options passed to asset_path will be passed to `compute_asset_path
```
Here you could write `+asset_path+` and `+compute_asset_path+`; if you don't wrap `asset_path` around pluses, RDoc will generate a link back to this method, which is pretty useless.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      # === Without the asset pipeline (`public_folder: true`)
+      #
+      # Accepts a `type` option that can specify the asset's extension. No error
+      # checking is done to verify the source passed into `asset_path` is valid
```
Ditto, you can use pluses here.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #
+      # === Without the asset pipeline (`public_folder: true`)
+      #
+      # Accepts a `type` option that can specify the asset's extension. No error
```
Ditto.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #
+      # === Options applying to all assets
+      #
+      # Below lists scenarios that apply to `asset_path` whether or not you're
```
Ditto.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #
+      #     asset_path("") # => ""
+      #
+      # - If `config.relative_url_root` is specified, all assets will have that
```
Ditto.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #     Rails.application.config.relative_url_root = "bar"
+      #     asset_path("foo.js", public_folder: true) # => "bar/foo.js"
+      #
+      # - A different asset host can be specified via `config.action_controller.asset_host`
```
Ditto.

robin850 commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #     Rails.application.config.action_controller.asset_host = "assets.example.com"
+      #     asset_path("foo.js", public_folder: true) # => "http://assets.example.com/foo.js"
+      #
+      # - An extension name can be specified manually with `extname`.
```
Ditto.

robin850 commented on guides/source/asset_pipeline.md:
````diff
@@ -724,7 +737,7 @@ If you have other manifests or individual stylesheets and JavaScript files to
 
 ```ruby
-Rails.application.config.assets.precompile += %w( admin.js admin.css )
+Rails.application.config.assets.precompile += ['admin.js', 'admin.css', 'swfObject.js']
````
Really nit-picky but why did you change the style here ? This is not consistent with other code snippets and the code generated by Rails by default. :blush:

robin850 commented on guides/source/configuring.md:
```diff
@@ -169,6 +169,8 @@ pipeline is enabled. It is set to `true` by default.
 
 * `config.assets.precompile` allows you to specify additional assets (other than `application.css` and `application.js`) which are to be precompiled when `rake assets:precompile` is run.
 
+* `config.assets.unknown_asset_fallback` allows you to modify the behavior of the asset pipeline when an asset is not in the pipeline with recent verisons of sprockets-rails.
```
Also nit-picky but to be consistent, you could say like above, "in the pipeline, if you use sprockets-rails 3.2.0 or newer".

robin850 commented:
This is looking very good ! :+1:

Just a nit-pick again regarding documentation though :smile: it looks like you didn't document the `:public_poster_folder` option for the `video_tag` method, this may be useful.

schneems committed:
Fix asset_pipeline.md miss-merge

schneems committed:
Address doc style comments from @robin850

vipulnsward commented on guides/source/asset_pipeline.md:
```diff
@@ -586,6 +586,19 @@ in your application are included in the `config.assets.precompile` list.
+
+If "asset fallback" is enabled then when an asset cannot be found the path will be
+output instead and no error raised.
+
```
We should mention what the default behaviour is here.

vipulnsward commented on guides/source/configuring.md:
```diff
@@ -169,6 +169,8 @@ pipeline is enabled. It is set to `true` by default.
 
 * `config.assets.precompile` allows you to specify additional assets (other than `application.css` and `application.js`) which are to be precompiled when `rake assets:precompile` is run.
 
+* `config.assets.unknown_asset_fallback` allows you to modify the behavior of the asset pipeline when an asset is not in the pipeline, if you use sprockets-rails 3.2.0 or newer.
```
Ditto, we should mention, `defaults to ...`

vipulnsward commented:
@schneems is there a way to test the request *does* send in requested public resource, I see the test cases for debug output that check something does not exist.

schneems committed:
Document public_poster_folder option

schneems committed:
Explicitly mention defaults.

schneems commented:
@vipulnsward there is one test with `application.js` with and without `public_folder: true`. Check out `test "{ public_folder: true } does not use the asset pipeline" do`

dhh commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      #
+      # Accepts a <tt>type</tt> option that can specify the asset's extension. No error
+      # checking is done to verify the source passed into +asset_path+ is valid
+      # and that the file exists on disk.
```
Not the biggest fan of `public_folder`. It implies that we're doing something special with this, when all we do is pass this directly through. I'd think something negative like `skip_pipeline: true` would be clearer for that. Would match this documentation as well.

schneems committed:
Address comment via @dhh, better option naming

kaspth commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -269,6 +270,8 @@ def image_alt(src)
       #   # => <video preload="none" controls="controls" src="/videos/trailer.ogg" ></video>
       #   video_tag("trailer.m4v", size: "16x10", poster: "screenshot.png")
       #   # => <video src="/videos/trailer.m4v" width="16" height="10" poster="/assets/screenshot.png"></video>
+      #   video_tag("trailer.m4v", size: "16x10", poster: "screenshot.png", public_poster_folder: true)
```
Shouldn't this just be `skip_pipeline` to match the other options? Presumably the same reasons renaming `public_folder` to `skip_pipeline` works here, so what do we gain making this stand out?

schneems replied:
There need to be 2 options here as the video may be in the pipeline but the poster might not be. Good catch though on the naming. I'll fix this.

kaspth replied:
Ah gotcha!

kaspth commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      # All options passed to +asset_path+ will be passed to +compute_asset_path+
+      # which is implemented by sprockets-rails.
+      #
+      #      asset_path("application.js") # => "/assets/application-60aa4fdc5cea14baf5400fba1abf4f2a46a5166bad4772b1effe341570f07de9.js"
```
I think our style is to nest code examples by 3 spaces, not 6.

schneems replied:
When you do that the example lines up with the bullet points, is that correct?

```
      # - An extension name can be specified manually with <tt>extname</tt>.
      #
      #   asset_path("foo", skip_pipeline: true, extname: ".js")     # => "/foo.js"
      #   asset_path("foo.css", skip_pipeline: true, extname: ".js") # => "/foo.css.js"
```

kaspth replied:
Ah, overlooked the bullet points! I guess the style then is 2 spaces from the starting indentation of a sentence. So here it would be 5 spaces.

kaspth replied:
In fact the examples seem to mix 6 and 5 spaces, so let's just have them all be 5 😁

schneems replied:
Some examples are bullet points and some aren't which is why they are different. Thanks for the reviews btw.

kaspth replied:
Then they should only have had 3 space indentation 😉

kaspth replied:
Ah sorry, they did get shortened to 3 spaces! GitHub somehow showed me an older diff.

kaspth commented on actionview/lib/action_view/helpers/asset_url_helper.rb:
```diff
@@ -118,16 +118,67 @@ module Helpers
+      # - An extension name can be specified manually with <tt>extname</tt>.
+      #
+      #     asset_path("foo", skip_pipeline: true, extname: ".js")     # => "/foo.js"
+      #     asset_path("foo.css", skip_pipeline: true, extname: ".js") # => "/foo.css.js"
```
Should these last 3 examples use `skip_pipeline`?

schneems replied:
Makes the example more clear IMHO

kaspth replied:
Yet I was confused by the usage 😄

I'm not objecting to using the option necessarily, but why does using it make the examples more clear?

schneems replied:
The assetpipeline does a bunch of things and I don't want asset pipeline magic being confused with this option. Adding a long fingerprint to the example makes it less clear that an extension was added to the output. What it doesn't do is replace an extension name, so `foo.css` does not become `foo.js`. Also technically this option doesn't change the extension output, it changes the input that goes into the asset pipeline. It's harder to show.

kaspth replied:
Okay, sounds good 👍

kaspth commented on railties/test/application/asset_debugging_test.rb:
```diff
@@ -68,5 +68,102 @@ class ::PostsController < ActionController::Base ; end
+      }
+
+      cases.each do |(view_method, tag_match)|
+        app_file "app/views/posts/index.html.erb", "<%= #{ view_method } '#{contents}', skip_pipeline: true %>"
```
Mixed spaces around interpolation braces abound in this test file. I think our preferred style is to not have any around interpolation braces, e.g.: `#{view_method}`.

kaspth commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -282,8 +285,11 @@ def image_alt(src)
       #   video_tag(["trailer.ogg", "trailer.flv"], size: "160x120")
       #   # => <video height="120" width="160"><source src="/videos/trailer.ogg" /><source src="/videos/trailer.flv" /></video>
       def video_tag(*sources)
```
We could remove the options wrangling with a keyword argument:

```ruby
def video_tag(*sources, skip_pipeline_on_poster: nil)
  multiple_sources_tag_builder("video", sources) do |options|
    options[:poster] = path_to_image(options[:poster], skip_pipeline: skip_pipeline_on_poster) if options[:poster]
    options[:width], options[:height] = extract_dimensions(options.delete(:size)) if options[:size]
  end
end
```

I think that would also save us a block argument shadowing local variable warning from `options`. 😁

schneems replied:
It's turtles all the way down, other method signatures that this gets passed to use extract_options.

Better to do that in a separate PR.

schneems replied:
Oh, i see what you're saying. I still want to do that in a different PR.

kaspth replied:
I'm not generally objecting to `extract_options!`, I agree that's out of scope. I'm objecting to using it in `video_tag` because the diff says it didn't use it before. But messing around in IRB showed me that we can't mix keyword arguments with a symbol keyed options hash without using `**options`, so that doesn't save us much here anyway. 👍 to keeping it like you had it 😁

schneems committed:
Better keyword argument name

schneems committed:
Fix style
